File: go/sorting/quick_sort.go
package sorting

type QuickSort struct{}

func (q QuickSort) Sort(arr []int, sortType SortType) []int {
	length := len(arr)
	return q.QuickSort(arr, 0, length-1, sortType)
}

func (q QuickSort) QuickSort(arr []int, low int, high int, sortType SortType) []int {
	if low < high {
		pivot := q.Partition(arr, low, high, sortType)
		q.QuickSort(arr, low, pivot-1, sortType)
		q.QuickSort(arr, pivot+1, high, sortType)
	}
	return arr
}

func (q QuickSort) Partition(arr []int, low int, high int, sortType SortType) int {
	pivot := arr[high]

	i := low - 1
	j := high

	for true {
		i++

		for Compare(pivot, arr[i], sortType) {
			i++
		}

		j--

		for i < j && Compare(arr[j], pivot, sortType) {
			j--
		}

		if j <= i {
			break
		}

		arr[i], arr[j] = arr[j], arr[i]
	}

	arr[i], arr[high] = arr[high], arr[i]

	return i
}

File: dart/lib/sorting/quick_sort.dart
import 'package:algorithm/sorting/sort.dart';

class QuickSort extends Sort {
  @override
  List<int> sort(List<int> arr, {SortType type = SortType.asc}) {
    final length = arr.length;

    return quickSort(arr, 0, length - 1, type);
  }

  int partition(List<int> arr, int low, int high, SortType type) {
    final pivot = arr[high];

    var i = low - 1;
    var j = high;

    while (true) {
      while (compare(pivot, arr[++i], type)) {}
      while (i < --j && compare(arr[j], pivot, type)) {}
      if (j <= i) break;

      arr.swap(i, j);
    }

    arr.swap(i, high);
    return i;
  }

  List<int> quickSort(List<int> arr, int low, int high, SortType type) {
    if (low < high) {
      final pivot = partition(arr, low, high, type);
      quickSort(arr, low, pivot - 1, type);
      quickSort(arr, pivot + 1, high, type);
    }
    return arr;
  }
}
